extensions/wikia/Flags/FlagsExtractor.class.php

```diff
@@ -114,10 +174,268 @@ public function getTemplate() {
+	 * @return null|string template name
+	 */
+	public function findFirstTemplateFromList( Array $templateNames, $text = null ) {
+		$minPosition = null;
+		$firstTemplate = null;
+
+		if ( !is_null( $text ) ) {
+			$this->text = $text;
+		}
+
+		foreach ( $templateNames as $templateName ) {
+			$templateBegin = "{{{$templateName}";
+			if ( ( $position = $this->findTemplatePosition( $templateBegin, 0 ) ) !== false ) {
+				if ( $position === 0 ) {
+					return $templateName;
+				}
+
+				if ( is_null( $minPosition ) || $position < $minPosition ) {
+					$minPosition = $position;
+					$firstTemplate = $templateName;
+				}
+			}
+		}
+
+		return $firstTemplate;
+	}
+
+	/**
+	 * Do some action after template is parsed
+	 *
+	 * Actions:
+	 *
+	 * - Add first template as flag (@see ACTION_ADD_FIRST_FLAG)
+	 * - Add all templates as flags (@see ACTION_ADD_ALL_FLAGS)
+	 * - Remove first template from text (@see ACTION_REMOVE_FIRST_FLAG)
+	 * - Remove all templates from text (@see ACTION_REMOVE_ALL_FLAGS)
+	 * - Replace first template by tag (@see ACTION_REPLACE_FIRST_FLAG)
+	 * - Replace all templates by tag (@see ACTION_REMOVE_ALL_FLAGS)
+	 *
+	 * Actions for first template have higher priority
+	 * For example:
+	 *
+	 * If we set "replace first template" and "remove all templates",
+	 * first template will be replaced, all others will be removed
+	 *
+	 * How to use:
+	 *
+	 * if we want: add first template as flag, replace first template and removed all others we need to pass array
+	 * [
+	 * 		self::ACTION_ADD_FIRST_FLAG,
+	 * 		self::ACTION_REPLACE_FIRST_FLAG,
+	 * 		self::ACTION_REMOVE_ALL_FLAGS
+	 * ]
+	 *
+	 * @return bool
+	 */
+	private function doTemplateActions() {
+		if ( empty( $this->actions ) ) {
+			return false;
+		}
+
+		$sum = array_sum( $this->actions );
+
+		$actionForFirstFlag = ( $sum & ( self::ACTION_REPLACE_FIRST_FLAG | self::ACTION_REMOVE_FIRST_FLAG ) ) && $this->isFirst;
+		$actionForAllFlags = $sum & ( self::ACTION_REMOVE_ALL_FLAGS | self::ACTION_REPLACE_ALL_FLAGS );
+
+		if ( ( $sum & ( self::ACTION_ADD_FIRST_FLAG ) && $this->isFirst )
+			|| ( $sum & self::ACTION_ADD_ALL_FLAGS ) ) {
+			$this->addFlagToPage();
+		}
+
+		if ( $actionForFirstFlag ) {
+			if ( $sum & self::ACTION_REPLACE_FIRST_FLAG ) {
+				$this->replaceTemplateByTag();
+			} else {
+				$this->removeTemplate();
+			}
+		} elseif ( $actionForAllFlags ) {
+			if ( $sum & self::ACTION_REPLACE_ALL_FLAGS ) {
+				$this->replaceTemplateByTag();
+			} else {
+				$this->removeTemplate();
+			}
+		}
+	}
+
+	/**
+	 * Add flag to page
+	 *
+	 * Parameters:
+	 * - wiki_id (required)
+	 * - page_id (required)
+	 * - flag_type_id (optional)
+	 *
+	 * @return bool
+	 */
+	private function addFlagToPage() {
+		if ( !empty( $this->actionParams )
+			&& isset( $this->actionParams['wiki_id'] )
+			&& isset( $this->actionParams['page_id'] )
+		) {
+			$flagsToPages = $this->prepareParameters();
+
+			if ( !$flagsToPages ) {
+				return false;
```

Comment:
This needs some sort of logging, even if it's just an echo to stdout. We'd need to know that this failed.

Up to you weather this would be an exception or just a log about the error